faculty/clients/template.py

```diff
@@ -42,4 +45,310 @@ def publish_new(self, project_id, template, source_directory):
             "sourceDirectory": source_directory,
             "name": template,
         }
-        return self._post_raw(endpoint, json=payload)
+        response = self._post_raw(endpoint, json=payload, check_status=False)
+        return _handle_publishing_response(response)
+
+    def add_to_project_from_directory(
+        self,
+        source_project_id,
+        source_directory,
+        target_project_id,
+        target_directory,
+        parameters,
+    ):
+        endpoint = "project/{}/apply".format(target_project_id)
+        payload = {
+            "sourceProjectId": str(source_project_id),
+            "sourceDirectory": source_directory,
+            "targetDirectory": target_directory,
+            "parameterValues": parameters,
+        }
+        response = self._post_raw(endpoint, json=payload, check_status=False)
+        return _handle_publishing_response(response)
+
+
+def _handle_publishing_response(response):
+    if 200 <= response.status_code < 300:
+        return response
+    elif 400 <= response.status_code < 500:
+        response_body = response.json()
+        error_code = response_body.get("errorCode")
+        if error_code == "resources_validation_failure":
+            raise ResourceValidationFailureResponseSchema().load(response_body)
+        elif error_code == "parameter_validation_failure":
+            raise ParameterValidationFailureSchema().load(response_body)
+        elif error_code == "template_retrieval_failure":
+            raise TemplateRetrievalFailureResponseSchema().load(response_body)
+        elif error_code == "default_parameters_parsing_error":
+            raise DefaultParametersParsingErrorSchema().load(response_body)
+        elif error_code == "generic_parsing_failure":
+            raise GenericParsingErrorSchema().load(response_body)
+        elif error_code == "workspace_files_validation_error":
+            raise WorkspaceFilesValidationErrorSchema().load(response_body)
+    else:
+        raise Exception(
```

Comment:
Rather than raising a vanilla `Exception`, it would be better if we could surface the custom exceptions that the library would normally raise. This could be done with a small tweak to your logic - instead of passing `check_status=False`, allow `_post_raw` to raise an exception, handle recognised cases, and re-raise any unhandled cases:

```
from faculty.clients.base import HttpError

...

    def add_to_project_from_directory(...):
        endpoint = "project/{}/apply".format(target_project_id)
        payload = {...}
        try:
            self._post_raw(endpoint, json=payload)
        except HttpError as err:
            _handle_publishing_error(err)

def _handle_publishing_error(exception):
    if exception.error_code == "resources_validation_failure":
        raise ResourceValidationFailureResponseSchema().load(exception.response.json())
    elif exception.error_code == "parameter_validation_failure":
        raise ParameterValidationFailureSchema().load(exception.response.json())
    elif exception.error_code == "template_retrieval_failure":
        raise TemplateRetrievalFailureResponseSchema().load(exception.response.json())
    elif exception.error_code == "default_parameters_parsing_error":
        raise DefaultParametersParsingErrorSchema().load(exception.response.json())
    elif exception.error_code == "generic_parsing_failure":
        raise GenericParsingErrorSchema().load(exception.response.json())
    elif exception.error_code == "workspace_files_validation_error":
        raise WorkspaceFilesValidationErrorSchema().load(exception.response.json())
    else:
        raise
```

Reply:
I thought this wasn't possible because I didn't realise the HTTP errors from the BaseClient still contain the original `response`. Good call, thanks.